### PR TITLE
Shipit v1<->v2 interaction and design

### PR DESCRIPTION
**Pull request by lundjordan**
Last week we met and discussed how shipit v1 and v2 would initially interact, at a deeper design level.

I've tried to convey what we agreed on via this PR.

@catlee @garbas and others, any concerns?

### CODE REVIEW

**Comment by lundjordan** [comment]
I'll of course squash before merging.

File: src/releng_docs/services_shipit.rst [hunk 1]
```diff
@@ -56,4 +56,63 @@ It does the following tasks on every run:
...
+    % curl -X POST -d "product=firefox&version=99.0" https://ship-it.mozilla.org/submit_release.html
+
+releaserunner_ continues to behave as normal by polling shipit v1 for new releases and, when found, runs release_sanity_, creates taskcluster_graph1_, and notifies release-drivers of new release.
+In addition, releaserunner will tell shipit-v2 about the new release:
+
+.. code-block:: python
+
+    r = requests.post("https://mozilla-releng.net/shipit_pipeline", data={uid='foo', pipeline={}})
+
+**shipit-v2 will create a pipeline of steps**
```
Comment (catlee) [inline src/releng_docs/services_shipit.rst]: would it be more accurate to say that release runner will be modified to create the pipeline, and shipit-v2 will manage it?

Comment (lundjordan) [inline src/releng_docs/services_shipit.rst]: I tried to convey that in code block above and the "releaserunner will tell shipit-v2 about the new release" part but I see what you mean and I'll fix this part

File: src/releng_docs/services_shipit.rst [hunk 2]
```diff
@@ -56,4 +56,61 @@ It does the following tasks on every run:
...
+.. code-block:: bash
+
+    % curl -X POST -d "product=firefox&version=99.0" https://ship-it.mozilla.org/submit_release.html
+
+releaserunner_ continues to behave as normal by polling shipit v1 for new releases and, when found, runs release_sanity_, creates taskcluster_graph1_, and notifies release-drivers of new release.
+In addition, releaserunner will tell shipit-v2 about the new release and ask shipit-v2 to create a pipeline:
+
+.. code-block:: python
+
+    r = requests.post("https://mozilla-releng.net/shipit_pipeline", data={uid='foo', pipeline={}})
```
Comment (garbas) [inline src/releng_docs/services_shipit.rst]: The URL for pipeline API service is going to be ``https://pipeline.shipit.mozilla-releng.net``.

This follows the naming convention of <service>.<domain> where <domain> is ``shipit.mozilla-releng.net`` (we might change this later if needed) and <service> is the name of the application in ``mozilla-releng/services`` without *family* name. eg: ``shipit_something`` has a *family* is ``shipit`` and name is ``something``. Family names are used to group development environments together, eg. provide one docker image for all of the *family*. Hope this clears things a bit.

File: src/releng_docs/services_shipit.rst [hunk 3]
```diff
@@ -56,4 +56,61 @@ It does the following tasks on every run:
...
+
+
+``shipit authentication and authorization``
+-------------------------------------------
+
+shipit-v2, at least on initial design, will adopt Taskcluster auth services for both authentication and authorization
+
+This means that rather than re-inventing the wheel or tracking user permissions, shipit-v2 will rely on Taskcluster for its client_, roles_, and scopes model when interacting with any shipit_* api and ui.
+
+As part of a hardened security effort, shipit will then implement MFA either by extending the Taskcluster auth service or adding third party MFA integration.
```
Comment (mozbhearsum) [inline src/releng_docs/services_shipit.rst]: Do you have a concrete idea of what it means to extend the Taskcluster auth service? For signoffs, we've been talking about storing OTPs secrets for each user within the signoffs service, which I'm not sure counts as "extending Taskcluster auth".

File: src/releng_docs/services_shipit.rst [hunk 4]
```diff
@@ -56,4 +56,61 @@ It does the following tasks on every run:
...
+    % curl -X POST -d "product=firefox&version=99.0" https://ship-it.mozilla.org/submit_release.html
+
+releaserunner_ continues to behave as normal by polling shipit v1 for new releases and, when found, runs release_sanity_, creates taskcluster_graph1_, and notifies release-drivers of new release.
+In addition, releaserunner will tell shipit-v2 about the new release and ask shipit-v2 to create a pipeline:
+
+.. code-block:: python
+
+    r = requests.post("https://mozilla-releng.net/shipit_pipeline", data={uid='foo', pipeline={}})
+
+shipit-v2's pipeline will consist of steps representing the release. A step could be sign off step (shipit_signoff), or a taskcluster step (shipit_taskcluster).
```
Comment (mozbhearsum) [inline src/releng_docs/services_shipit.rst]: This reads as if shipit_signoff and shipit_taskcluster are the only available steps. Rewording this slightly might make it hold up better over time.

File: src/releng_docs/services_shipit.rst [hunk 5]
```diff
@@ -56,4 +56,61 @@ It does the following tasks on every run:
...
+
+**Betas**
+
+.. image:: img/beta_overview.jpg
+
+In the beta case, graph1 would finish with 'push artifacts to release location'. There would then be a shipit-v2 pipeline consisting of a sign off step: 'publish release human sign off', and a taskcluster step: 'publish release'
+
+The taskcluster 'publish release' step would comprise of the following tasks: publishing on balrog, bumping next version, updating bouncer, and informing shipit-v1 the release is complete (mark as shipped).
+
+Release candidates would be implemented in a similar manner but since it contains more sign offs, shipit-v1 would offload more tasks to shipit-v2
```
Comment (mozbhearsum) [inline src/releng_docs/services_shipit.rst]: This surprise me a little. I had it in my head that graph1 would stay exactly as is for our initial implementation, and that we'd start moving parts of it to ship it v2 afterwards. Even in this scenario, ship it v2 tracks the final signoff before we ship, so we gain the benefit of formalizing that in the signoffs service.

I don't feel strongly about this either way, just trying to clarify - please correct me if I'm wrong!

File: src/releng_docs/services_shipit.rst (hunk 3 above)
Comment (garbas) [inline src/releng_docs/services_shipit.rst]: Extending of taskcluster auth can happen both (1) within the existing services if we reach the consensus with TC team or (2) we create a new service which is extending TC auth with things that we are missing from it.

File: src/releng_docs/services_shipit.rst [hunk 6]
```diff
@@ -56,4 +56,61 @@ It does the following tasks on every run:
...
+
+
+``shipit authentication and authorization``
+-------------------------------------------
+
+shipit-v2, at least on initial design, will adopt Taskcluster auth services for both authentication and authorization
+
+This means that rather than re-inventing the wheel or tracking user permissions, shipit-v2 will rely on Taskcluster for its client_, roles_, and scopes model when interacting with any shipit_* api and ui.
```
Comment (garbas) [inline src/releng_docs/services_shipit.rst]: I think at one of the last meetings I was present there was no real argument against it.

Comment (mozbhearsum) [inline src/releng_docs/services_shipit.rst]: We just talked about it again today, and it looks like we are!

File: src/releng_docs/services_shipit.rst [hunk 7]
```diff
@@ -56,4 +56,61 @@ It does the following tasks on every run:
...
+
+Since graph1 is created by shipit-v1 and includes taskcluster tasks for doing sign offs and publishing releases, it will need to be trimmed and then offload its later tasks to shipit-v2 via subsequent steps within the pipeline.
+
+And so, initially, graph1 will create all required tasks up until the first human sign off. This is easily defined by: 'the first "human decision task" within the graph'.
+
+For example, here is an overview of what a Beta release comprises of in simplified form:
+
+**Betas**
+
+.. image:: img/beta_overview.jpg
```
Comment (lundjordan) [inline src/releng_docs/services_shipit.rst]: true, I should but I will defer that to the next time it needs to be updated as it's done now. is that okay?

File: src/releng_docs/services_shipit.rst (hunk 5 above)
Comment (lundjordan) [inline src/releng_docs/services_shipit.rst]: hm, you may be correct. My understanding was we intended to leave graph1 as is. However, beta's currently only have one graph. So graph1 handles all the sign offs (human decision tasks) and even publishing on balrog + postrelease tasks.

Therefore, v2 would be left with nothing to do. At subsequent meetings I suggested we break graph1 at current human decision tasks. Each human task is translated to a signoff step, and we have a tc graph (step) in between each sign off (step).

Given that, do you still see differently?

File: src/releng_docs/services_shipit.rst (hunk 6 above)
Comment (lundjordan) [inline src/releng_docs/services_shipit.rst]: given our recent auth0 and tc auth discussions, I'm going to defer defining this for now. As temp docs, I could put the three options we finalized on, if you haven't written that down somewhere yet? @garbas @mozbhearsum ?

File: src/releng_docs/services_shipit.rst (hunk 5 above)
Comment (lundjordan) [inline src/releng_docs/services_shipit.rst]: @mozbhearsum ^

**Comment by lundjordan** [comment]
squashed and separated by two commits.

Former is original w/ some bitrot since: https://github.com/mozilla-releng/services/commit/29d361825c96b0b43c585ff65ad0140134b63ed4

The latter highlights changes as per feedback in this PR. I added a quick summary of what I interpreted as the agreed auth solutions agreed upon from Friday's shipit meeting.

File: src/releng_docs/services_shipit.rst (hunk 7 above)
Comment (garbas) [inline src/releng_docs/services_shipit.rst]: I've added a PR against your branch, https://github.com/lundjordan/services/pull/1, for this change

**Comment by lundjordan** [comment]
okay, fixed some bitrot around the auth options, pulled in garbas blockdiag, and merged with catlee's pipeline docs.

anybody want to do final sign off? @garbas @mozbhearsum ?

File: src/releng_docs/run.py [hunk 8]
```diff
@@ -5,6 +5,7 @@
 
 HERE = os.path.dirname(__file__)
 
+os.system('/run/current-system/sw/bin/make html')
```
Comment (catlee) [inline src/releng_docs/run.py]: what does this do?

Comment (lundjordan) [inline src/releng_docs/run.py]: pulled from garbas PR.

@garbas ?

Comment (lundjordan) [inline src/releng_docs/run.py]: https://github.com/mozilla-releng/services/pull/226/commits/264c160f241133de9b0be83903b71b8528744d5d

Comment (garbas) [inline src/releng_docs/run.py]: i think i fixed this in the master already. @lundjordan can you rebase on ``origin/master``?

**Comment by garbas** [comment]
I rebased this work on current master and will push it as it is. Improvements to ShipIt documentation will follow later.

**Pull request merged**